From 3d7cc461d78451cda566d6994a30ae8e1e789575 Mon Sep 17 00:00:00 2001
From: Eric Blake <eblake@redhat.com>
Date: Sat, 31 May 2025 07:37:28 -0500
Subject: [PATCH] generator: Avoid const-correctness warnings in golang

Hack the generator to add the necessary casts to discard const in a
way that shuts up the warnings from compiling wrappers.go.

Signed-off-by: Eric Blake <eblake@redhat.com>
(cherry picked from commit a909e74f902bb9d1e8a4ab87ae5ccf76d4675787)
---
 generator/C.ml      | 10 ++++++----
 generator/C.mli     |  2 +-
 generator/GoLang.ml |  4 ++--
 3 files changed, 9 insertions(+), 7 deletions(-)

diff --git a/generator/C.ml b/generator/C.ml
index ad08437c..50d22306 100644
--- a/generator/C.ml
+++ b/generator/C.ml
@@ -286,16 +286,16 @@ let print_fndecl ?wrap ?closure_style name args optargs ret =
   pr "\n    LIBNBD_ATTRIBUTE_NONNULL (%s);\n" (String.concat ", " nns)
 
 let rec print_cbarg_list ?(wrap = false) ?maxcol ?types ?(parens = true)
-          cbargs =
+          ?(noconst = false) cbargs =
   if parens then pr "(";
   if wrap then
     pr_wrap ?maxcol ','
-      (fun () -> print_cbarg_list' ?types cbargs)
+      (fun () -> print_cbarg_list' ?types noconst cbargs)
   else
-    print_cbarg_list' ?types cbargs;
+    print_cbarg_list' ?types noconst cbargs;
   if parens then pr ")"
 
-and print_cbarg_list' ?(types = true) cbargs =
+and print_cbarg_list' ?(types = true) noconst cbargs =
   if types then pr "void *";
   pr "user_data";
 
@@ -316,6 +316,7 @@ and print_cbarg_list' ?(types = true) cbargs =
       | CBArrayAndLen _ -> assert false
       | CBBytesIn (n, len) ->
          if types then pr "const void *";
+         if noconst then pr "(void *)";
          pr "%s, " n;
          if types then pr "size_t ";
          pr "%s" len
@@ -331,6 +332,7 @@ and print_cbarg_list' ?(types = true) cbargs =
       | CBMutable arg -> assert false
       | CBString n ->
          if types then pr "const char *";
+         if noconst then pr "(char *)";
          pr "%s" n
       | CBUInt n ->
          if types then pr "unsigned ";
diff --git a/generator/C.mli b/generator/C.mli
index a4b31351..75d77276 100644
--- a/generator/C.mli
+++ b/generator/C.mli
@@ -34,7 +34,7 @@ val print_arg_list : ?wrap:bool -> ?maxcol:int ->
                      ?closure_style:closure_style ->
                      API.arg list -> API.optarg list -> unit
 val print_cbarg_list : ?wrap:bool -> ?maxcol:int ->
-                       ?types:bool -> ?parens:bool ->
+                       ?types:bool -> ?parens:bool -> ?noconst:bool ->
                        API.cbarg list -> unit
 val print_call : ?wrap:bool -> ?maxcol:int ->
                  ?closure_style:closure_style ->
diff --git a/generator/GoLang.ml b/generator/GoLang.ml
index 3fe7cd53..1505a598 100644
--- a/generator/GoLang.ml
+++ b/generator/GoLang.ml
@@ -159,9 +159,9 @@ let print_callback_wrapper { cbname; cbargs } =
   C.print_cbarg_list ~wrap:true cbargs;
   pr "\n";
   pr "{\n";
-  pr "  // golang isn't const-correct, there will be warnings here:\n";
+  pr "  // golang isn't const-correct, casts avoid warnings here:\n";
   pr "  return %s_callback ((long *)" cbname;
-  C.print_cbarg_list ~types:false ~parens:false cbargs;
+  C.print_cbarg_list ~types:false ~parens:false ~noconst:true cbargs;
   pr ");\n";
   pr "}\n";
   pr "\n";
-- 
2.47.3

